Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+- Hetzner load balancer: [https://github.com/syself/hetzner-cloud-controller-manager/blob/master/internal/annotation/load_balancer.go](https://github.com/syself/hetzner-cloud-controller-manager/blob/master/internal/annotation/load_balancer.go)
+
+This can be avoided rather easily. We have to implement a new interface for load balancers, as a counterpart to the current one ([https://pkg.go.dev/k8s.io/cloud-provider#LoadBalancer](https://pkg.go.dev/k8s.io/cloud-provider#LoadBalancer)), which is a part of the cloud provider interface ([https://pkg.go.dev/k8s.io/cloud-provider#Interface](https://pkg.go.dev/k8s.io/cloud-provider#Interface)). SCS implements the current cloud provider interface and defines a new one, which SCS cloud providers have to implement. In there, we unify the naming of the annotations and give a pre-defined set of inputs to the function that, for example, creates a load balancer.
+
+To be more precise: Instead of `"load-balancer.hetzner.cloud/id"` and `"[loadbalancer.openstack.org/load-balancer-id](http://loadbalancer.openstack.org/load-balancer-id)"` we would establish a standard that could be named `"load-balancer.scs.cloud/id"`. The SCS load balancer controller interface knows these annotations and converts them into a function like the following:
+ `func CreateLoadbalancer(id string, otherInputs object) error`
+
+Note that the ID here comes from the aforementioned annotation. Of course, this function will have more parameters after the standardizing process is completed. `otherInputs` serves as a placeholder for all those still missing inputs. To define precisely the annotations and parameters of the functions that cloud providers need to implement, further discussion is required in the SCS team.
+
+Note that standardizing the inputs has the effect that cloud providers cannot have certain individual features, which require some unique inputs.
+
+The second task is to standardize the naming of storage classes.
+
+Different cloud providers have different storage classes and name them differently. As these names of the storage classes are used in persistent volume claims (i.e. parts of the configuration of our applications) ([https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reserving-a-persistentvolume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reserving-a-persistentvolume)), we would need a separate configuration for each application and cloud provider. Therefore, there wouldn’t be any standardization. Instead, we propose to standardize the naming of storage classes, so that the PVCs in the configuration of applications can be defined independently of the provider.
+
+### Drawbacks
+
+Standardizing something always leads to less freedom of choice. Cloud providers would not be able to have individual inputs for creating load balancers, so some of their features might be obsolete.
```

Comment:
This can be mitigated by using only a mutating hook instead of a custom object; then additional annotations could simply be passed through.